src/_P001_Switch.ino

```diff
@@ -173,51 +244,152 @@ boolean Plugin_001(byte function, struct EventStruct *event, String& string)
     case PLUGIN_TEN_PER_SECOND:
       {
         const boolean state = Plugin_001_read_switch_state(event);
+
+        /**************************************************************************\
+        20181009 - @giig1967g: new doubleclick logic is:
+        if there is a 'state' change, check debounce period.
+        Then if doubleclick interval exceeded, reset Plugin_001_clickCounter to 0
+        Plugin_001_clickCounter contains the current status for doubleclick:
+        0: start counting
+        1: 1st click
+        2: 2nd click
+        3: 3rd click = doubleclick event if inside interval (calculated as: '3rd click time' minus '1st click time')
+
+        Returned EVENT value is = 3 always for doubleclick
+        In rules this can be checked:
+        on Button#Switch=3 do //will fire if doubleclick
+        \**************************************************************************/
+
```

Comment:
I see no differences between original plugin and this implementation.

Infact as you can see the code is huge, but it is divided into two sections with the `if` clause:
the first section is the original section and is executed ONLY when there is a state change. So very rarely. The second section is run only during the interval of the longpress action, so if a longpress interval is 2 seconds, it will run only 20 times and only during the actual seconds when the button is pressed.

The idea behind the event states is the following. I tried to keep the original logic, that is: a switch will return a number. Today it will 0 or 1 whe there is a state change. 0 is when the switch state changes to HIGH and 1 when the state changes to LOW.
As the state=2 is used for INPUT switches, I used state=3 to indicate a doubleclick.
So the rules will remain with the same logic:

```
on button#switch do
if [button#switch]=0 do 
//code if button is high
elseif [button#switch]=1 do 
//code if button is low
elseif [button#switch]=3 do 
//code if button is doubleclick
elseif [button#switch]=10 do 
//code if button is longpress high
elseif [button#switch]=11 do 
//code if button is longpress low
endif
```

Regarding the speed, I have set in my tests a de-bounce of 100-150ms and a doubleclick interval of 1000 ms. It is extremely reliable. Not a single failure.

> Why not add extra values to the event to indicate multi click? And one for long press?
> Then the event for Switch will remain the same as it is now.
Because it's not necessary and will change the coding logic.
Imagine the same logic in Windows: when you press the mouse button once you get a CLICK event. When you press it twice, you get a CLICK event and soon after you get a DOUBLECLICK event. I used the same logic. If you press once you get 0 (or 1), when you press twice, you get 0 (or 1) and 3.